refactor: make LndClient Supervisor friendly

ramontayag opened the pull request:
One can pass something like this to a Supervisor:

```ex
{
  LndClient,
  %LndClient.ConnConfig{
    node_uri: System.get_env("NODE_URI"),
    cert_path: System.get_env("CERT_PATH"),
    macaroon_path: System.get_env("MACAROON_PATH")
  },
},
```

This addresses https://github.com/RooSoft/lnd_client/issues/10, but:
- not the comment about restarting when it is disconnected -- that is to follow!
- not naming the genserver in the scenario that one would need to connect to multiple

Tested locally with Polar.

ramontayag committed:
refactor: make LndClient Supervisor friendly

One can pass something like this to a Supervisor:

```ex
{
  LndClient,
  %LndClient.ConnConfig{
    node_uri: System.get_env("NODE_URI"),
    cert_path: System.get_env("CERT_PATH"),
    macaroon_path: System.get_env("MACAROON_PATH")
  },
},
```

RooSoft commented:
Been skimming through the code, and it makes a lot of sense... let me try it.

RooSoft commented on lib/lnd_client.ex:
```diff
@@ -18,12 +21,41 @@ defmodule LndClient do
+  def start_link(%ConnConfig{} = conn_config) do
+    case GenServer.start_link(__MODULE__, init_state(conn_config), name: __MODULE__) do
+      {:ok, pid} = result ->
+        connect
```
`warning`: variable "connect" does not exist and is being expanded to "connect()", please use parentheses to remove the ambiguity or change the variable name

ramontayag replied:
Done. Will push soon. I didn't get this warning. How do I turn this on?

RooSoft commented on lib/lnd_client.ex:
```diff
@@ -18,12 +21,41 @@ defmodule LndClient do
+  """
+  def start_link(%ConnConfig{} = conn_config) do
+    case GenServer.start_link(__MODULE__, init_state(conn_config), name: __MODULE__) do
+      {:ok, pid} = result ->
```
as the `pid` is unused, rename it to `_pid` so this warning goes away

`warning`: variable "pid" is unused (if the variable is not meant to be used, prefix it with an underscore)

RooSoft commented on lib/lnd_client.ex:
````diff
@@ -18,12 +21,41 @@ defmodule LndClient do
+  ```
+  """
+  def start_link(%ConnConfig{} = conn_config) do
+    case GenServer.start_link(__MODULE__, init_state(conn_config), name: __MODULE__) do
````
I think it would make more sense to start the connection inside the GenServer as opposed to the current thread... thus, I'd remove that whole case statement around the `GenServer.start_link` and move it in `init`.

RooSoft replied:
I just noticed that the connect function makes a call to the GenServer before it's even been initiated...

I think the connect function should be a normal function, but called from within the `init` function, and thus in the context of the GenServer itself.

ramontayag replied:
Ah yes, that's right! Because in the init the server would have started already

ramontayag replied:
I can't seem to call `connect()` from the `init` function. Is it because it hasn't been started yet and therefore this exception is raised?

```
17:00:15.371 [error] Process #PID<0.77.0> raised an exception
** (ArgumentError) errors were found at the given arguments:

  * 2nd argument: not a pid

    :erlang.monitor(:process, {:error, {:calling_self, {GenServer, :call, [LndClient, :connect, :infinity]}}})
    (iex 1.14.0) lib/iex/server.ex:100: IEx.Server.run_without_registration/1
```

ramontayag replied:
Ah, well, scratch that. I got something working. What worked was to not make a call to the genserver, but rather connect already and return that state in `init`.

```ex
  def init(state) do
    connect(state) # call the Connectivity.connect code here
  end
```

The thing is this may return `{:error, ...}`. I haven't seen `init` ever return those in examples online.

RooSoft replied:
Yeah, `start_link` is in the process of the caller, but `init` is executing in the GenServer's own process. When calling from outside, you got to switch context, thus make calls or casts, but from init, all you got to do is to do the job straight away.

RooSoft commented on lib/lnd_client.ex:
```diff
@@ -186,8 +214,10 @@ defmodule LndClient do
+  def handle_call(:connect, _from, state) do
+    conn_config = Map.get(state, :conn_config)
+
+    case Connectivity.connect(conn_config |> Map.from_struct) do
```
I'd rather `LndClient.Connectivity` to receive a `%ConnConfig{}` rather than doing a `Map.from_struct` from here

RooSoft reviewed (changes requested):
This is a very welcome upgrade, thank you @ramontayag.

Just a couple of minor changes needed before the merge.

ramontayag committed:
refactor: LndClient.Connectivity accepts ConnConfig

ramontayag committed:
chore add underscore to pid since it is not used

ramontayag committed:
refactor: connect to LND upon start up

No need to make a call from another process to
the genserver to trigger a connection. We will
always want the genserver to attempt to connect
to LND when it starts.

RooSoft commented:
All good, thank you @ramontayag

ramontayag commented:
Thanks!
